File: Leetcode/3085 - Minimum Deletions to Make String K-Special/solution.ts
function minimumDeletions(word: string, k: number): number {
    const charCount: number[] = new Array(26).fill(0);

    for (let i: number = 0; i < word.length; i++) {
        const index: number = word.charCodeAt(i) - 97;
        charCount[index]++;
    }

    const freqArr: number[] = charCount.filter(count => count > 0);
    let minDeletions: number = word.length;

    for (let i: number = 0; i < freqArr.length; i++) {
        const base: number = freqArr[i];
        let deletions: number = 0;
        for (let j: number = 0; j < freqArr.length; j++) {
            const curr: number = freqArr[j];
            if (curr < base) {
                deletions += curr;
            } else if (curr > base + k) {
                deletions += curr - (base + k);
            }
        }
        minDeletions = Math.min(minDeletions, deletions);
    }

    return minDeletions;
};

File: Leetcode/3085 - Minimum Deletions to Make String K-Special/solution.py
def minimumDeletions(self, word: str, k: int) -> int:
    char_count: List[int] = [0] * 26
    for ch in word:
        index = ord(ch) - ord('a')
        char_count[index] += 1

    frequencies: List[int] = [count for count in char_count if count > 0]
    min_deletions: int = len(word)
    for base_freq in frequencies:
        deletions: int = 0
        for freq in frequencies:
            if freq < base_freq:
                deletions += freq
            elif freq > base_freq + k:
                deletions += freq - (base_freq + k)
        min_deletions = min(min_deletions, deletions)

    return min_deletions